applications/CompressiblePotentialFlowApplication/python_scripts/stochastic_potential_flow_response.py

```diff
@@ -0,0 +1,300 @@
+    model_part_name = solver_settings["model_part_name"].GetString()
+    if not model.HasModelPart(model_part_name):
+        model_part = model.CreateModelPart(model_part_name, 2)
+        domain_size = solver_settings["domain_size"].GetInt()
+        if domain_size < 0:
+            raise Exception('Please specify a "domain_size" >= 0!')
+        model_part.ProcessInfo.SetValue(KratosMultiphysics.DOMAIN_SIZE, domain_size)
+    else:
+        model_part = model.GetModelPart(model_part_name)
+
+    return model_part
+
+class AdjointResponseFunction(ResponseFunctionInterface):
+
+    def __init__(self, identifier, response_settings, model):
+        self.identifier = identifier
+        self.response_settings = response_settings
+        self.xmc_settings_path = response_settings["xmc_settings"].GetString()
+        self.design_surface_sub_model_part_name = response_settings["design_surface_sub_model_part_name"].GetString()
+        self.auxiliary_mdpa_path = response_settings["auxiliary_mdpa_path"].GetString()
+        # Create the primal solver
+        with open(self.response_settings["primal_settings"].GetString(),'r') as parameter_file:
+            primal_parameters = Parameters( parameter_file.read() )
+
+        primal_parameters = _CheckParameters(primal_parameters)
+        if primal_parameters.Has("adjoint_parameters_path"):
+            primal_parameters["adjoint_parameters_path"].SetString(self.response_settings["adjoint_settings"].GetString())
+        else:
+            primal_parameters.AddString("adjoint_parameters_path", self.response_settings["adjoint_settings"].GetString())
+        if primal_parameters.Has("design_surface_sub_model_part_name"):
+            primal_parameters["design_surface_sub_model_part_name"].SetString(self.design_surface_sub_model_part_name)
+        else:
+            primal_parameters.AddString("design_surface_sub_model_part_name", self.design_surface_sub_model_part_name)
+        open(self.response_settings["primal_settings"].GetString(), 'w').write(primal_parameters.PrettyPrintJsonString())
+
+        # Store current design
+        self.current_model_part = _GetModelPart(model, primal_parameters["solver_settings"])
+
+    def InitializeSolutionStep(self):
+        self.current_model_part.RemoveSubModelPart("fluid_computational_model_part")
+        KratosMultiphysics.ModelPartIO(self.auxiliary_mdpa_path, KratosMultiphysics.IO.WRITE | KratosMultiphysics.IO.MESH_ONLY).WriteModelPart( self.current_model_part)
+
+        self._RunXMC()
+
+        # TODO: here we should retrieve the risk measure we want to use for optimization
+        #       at the moment we are using the expected value estimation
+        risk_measure = "expected_value"
+        if risk_measure == "expected_value":
+            order = 1 ; is_central = False
+        elif risk_measure == "variance":
+            order = 2 ; is_central = True
+
+        # save lift coefficient
+        qoi_counter = 0
```

Comment:
If you want to add more QoIs this value should not be hardcoded bu should be updated within a loop. Up to you.

I have moved from a loop to hardcoded since you were only using `qoi_counter=0`.